casper/src/test/scala/coop/rchain/casper/util/CasperUtilTest.scala

```diff
@@ -0,0 +1,71 @@
+package coop.rchain.casper.util
+
+import com.google.protobuf.ByteString
+import coop.rchain.casper.protocol._
+import coop.rchain.casper.protocol.Resource.ResourceClass.ProduceResource
+import org.scalatest.{FlatSpec, Matchers}
+import coop.rchain.crypto.hash.Sha256
+
+import scala.collection.mutable
+
+class CasperUtilTest extends FlatSpec with Matchers {
+  var resourceCounter = 0
```

Comment:
@goral09 @rabbitonweb @birchmd Is there a good pattern to do this counter functionally?

Reply:
Basically I want to bump it up whenever createBlock is called and use the value within createBlock.

Reply:
Sounds like a `State` monad where the `S` (state) would be the `resourceCounter` producing entity and the value returned is `BlockMessage`. 

Or if you want something simpler, you can go with something more OO : 
```
class State() {
  var resourceCounter = 0
  def createBlock(parentsHashList: Seq[ByteString]) = {  
    // bump resourceCounter
    // call your logic
   }
}

Reply:
A marginally better way to do it (in my opinion), though not functional per se because it relies on side effects would be to use an iterator. 
```
  val recourceCounter = Iterator.from(0)
  ...
  private def createBlock(parentsHashList: Seq[ByteString]) = {
    val resourceId = resourceCounter.next()
    val uniqueResource = Resource(ProduceResource(Produce(resourceId)))
    ...
  }
```

Reply:
@KentShikama 

Here is very simple example using `State` monad:

```
type Hash = Int
case class Block(prevHashes: Seq[Hash], hash: Hash)
def createBlock(parentHashes: Seq[Hash]): State[Hash, Block] =
  State[Hash, Block](s => (s + 1, Block(parentHashes, s + 1)))

val s = for {
  genesis <- createBlock(Seq.empty)
  block1 <- createBlock(Seq(genesis.hash))
  block2 <- createBlock(Seq(block1.hash))
  block3 <- createBlock(Seq(block1.hash, block2.hash))
  // you can throw `assert` here if you want
  block4 <- createBlock(Seq(block1.hash, block3.hash))
} yield block4

s.runA(0).value
// res0: Block = Block(List(2, 4),5)
```

Reply:
@goral09 I've given a shot at using the State monad but I can't find a way to "leak" the individual blocks out.

```
    val blocks = for {
      genesis <- createBlock(Seq.empty)
      blockMessage2 <- createBlock(Seq(genesis.blockHash))
      blockMessage3 <- createBlock(Seq(genesis.blockHash))
      blockMessage4 <- createBlock(Seq(blockMessage2.blockHash, blockMessage3.blockHash))
    } yield blockMessage4

    val chain = blocks.runS(immutable.HashMap[ByteString, BlockMessage]()).value

    isInMainChain(chain, genesis, blockMessage2) should be(true) // This does not work as genesis or blockMessage2 is not in scope
    isInMainChain(chain, genesis, blockMessage3) should be(true)
    isInMainChain(chain, genesis, blockMessage4) should be(true)
    isInMainChain(chain, blockMessage2, blockMessage4) should be(true)
    isInMainChain(chain, blockMessage3, blockMessage4) should be(false)
```

and createBlock is defined as

```
  private def createBlock(parentsHashList: Seq[ByteString]): State[immutable.HashMap[ByteString, BlockMessage], BlockMessage] =
  State[immutable.HashMap[ByteString, BlockMessage], BlockMessage] {
    chain => {
      val uniqueResource = Resource(ProduceResource(Produce(chain.size)))
      val postState = RChainState().withResources(Seq(uniqueResource))
      val postStateHash = Sha256.hash(postState.toByteArray)
      val header = Header()
        .withPostStateHash(ByteString.copyFrom(postStateHash))
        .withParentsHashList(parentsHashList)
      val blockHash = Sha256.hash(header.toByteArray)
      val body = Body().withPostState(postState)
      val blockMessage = BlockMessage(ByteString.copyFrom(blockHash), Some(header), Some(body))
      (chain + (blockMessage.blockHash -> blockMessage), blockMessage)
    }
  }
```

Reply:
I'll go with the less functional approach that @birchmd mentioned for now but I'm really interested in seeing how this would have been done otherwise.